cmd/cloud-controller-manager/main.go

```diff
@@ -20,27 +20,70 @@ limitations under the License.
 package main
 
 import (
-	"k8s.io/cloud-provider-gcp/cmd/cloud-controller-manager/app"
 	"math/rand"
 	"os"
 	"time"
 
+	"github.com/spf13/pflag"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/cloud-provider"
+	"k8s.io/cloud-provider/app"
+	"k8s.io/cloud-provider/app/config"
+	"k8s.io/cloud-provider/options"
+	cliflag "k8s.io/component-base/cli/flag"
 	"k8s.io/component-base/logs"
-	_ "k8s.io/component-base/metrics/prometheus/version" // for version metric registration
-	// NOTE: Importing all in-tree cloud-providers is not required when
-	// implementing an out-of-tree cloud-provider.
-	_ "k8s.io/cloud-provider-gcp/providers/gce"
 	_ "k8s.io/component-base/metrics/prometheus/clientgo" // load all the prometheus client-go plugins
+	_ "k8s.io/component-base/metrics/prometheus/version"  // for version metric registration
+	"k8s.io/klog/v2"
+
+	_ "k8s.io/cloud-provider-gcp/providers/gce"
+	// For existing cloud providers, the option to import legacy providers is still available.
+	// e.g. _"k8s.io/legacy-cloud-providers/<provider>"
 )
 
 func main() {
 	rand.Seed(time.Now().UnixNano())
+
+	ccmOptions, err := options.NewCloudControllerManagerOptions()
+	if err != nil {
+		klog.Fatalf("unable to initialize command options: %v", err)
+	}
+
+	fss := cliflag.NamedFlagSets{}
+	command := app.NewCloudControllerManagerCommand(ccmOptions, cloudInitializer, app.DefaultInitFuncConstructors, fss, wait.NeverStop)
+
+	// TODO: once we switch everything over to Cobra commands, we can go back to calling
+	// utilflag.InitFlags() (by removing its pflag.Parse() call). For now, we have to set the
+	// normalize func and add the go flag set by hand.
+	// Here is an sample
+	pflag.CommandLine.SetNormalizeFunc(cliflag.WordSepNormalizeFunc)
+	// utilflag.InitFlags()
 	logs.InitLogs()
 	defer logs.FlushLogs()
 
-	command := app.NewCloudControllerManagerCommand()
-
 	if err := command.Execute(); err != nil {
 		os.Exit(1)
 	}
 }
+
+func cloudInitializer(config *config.CompletedConfig) cloudprovider.Interface {
+	cloudConfig := config.ComponentConfig.KubeCloudShared.CloudProvider
+
+	// initialize cloud provider with the cloud provider name and config file provided
+	cloud, err := cloudprovider.InitCloudProvider(cloudConfig.Name, cloudConfig.CloudConfigFile)
+	if err != nil {
+		klog.Fatalf("Cloud provider could not be initialized: %v", err)
```

Comment:
Please add cloudConfig.Name and cloudConfig.CloudConfigFile to the outputted fatal message.

Reply:
Updated